common/logging/EventAuditLogger.php

```diff
@@ -0,0 +1,1080 @@
+            $sslopts['cafile'] = $cafile;
+            $sslopts['verify_peer'] = true;
+            $sslopts['verify_depth'] = 10;
+        }
+
+        if ($localcert !== null && $localcert != "") {
+            $sslopts['local_cert'] = $localcert;
+        }
+
+        $opts = array('tls' => $sslopts, 'ssl' => $sslopts);
+        $ctx = stream_context_create($opts);
+        $timeout = 60;
+        $flags = STREAM_CLIENT_CONNECT;
+
+        $olderr = error_reporting(0);
+        $conn = stream_socket_client(
+            'tls://' . $host . ":" . $port,
+            $errno,
+            $errstr,
+            $timeout,
+            $flags,
+            $ctx
+        );
+        error_reporting($olderr);
+        return $conn;
+    }
+
+    /**
+     * This function is used to send audit records to an Audit Repository Server,
+     * as described in the Audit Trail and Node Authentication (ATNA) standard.
+     * Given the fields in a single audit record:
+     * - Create an XML audit message according to RFC 3881, including the RFC5425 syslog header.
+     * - Create a TLS connection that performs bi-directions certificate authentication,
+     *   according to RFC 5425.
+     * - Send the XML message on the TLS connection.
+     *
+     * @param $user
+     * @param $group
+     * @param $event
+     * @param $patient_id
+     * @param $outcome
+     * @param $comments
+     */
+    public function send_atna_audit_msg($user, $group, $event, $patient_id, $outcome, $comments)
+    {
+        /* If no ATNA repository server is configured, return */
+        if (empty($GLOBALS['atna_audit_host']) || empty($GLOBALS['enable_atna_audit'])) {
+            return;
+        }
+
+        $host = $GLOBALS['atna_audit_host'];
+        $port = $GLOBALS['atna_audit_port'];
+        $localcert = $GLOBALS['atna_audit_localcert'];
+        $cacert = $GLOBALS['atna_audit_cacert'];
+        $conn = $this->create_tls_conn($host, $port, $localcert, $cacert);
+        if ($conn !== false) {
+            $msg = $this->create_rfc3881_msg($user, $group, $event, $patient_id, $outcome, $comments);
+            fwrite($conn, $msg);
+            fclose($conn);
+        }
+    }
+
+    /**
+     * Add an entry into the audit log table, indicating that an
+     * SQL query was performed. $outcome is true if the statement
+     * successfully completed.  Determine the event type based on
+     * the tables present in the SQL query.
+     *
+     * @param $statement
+     * @param $outcome
+     * @param null      $binds
+     */
+    public function auditSQLEvent($statement, $outcome, $binds = null)
+    {
+        $user =  $_SESSION['authUser'] ?? "";
+
+        /* Don't log anything if the audit logging is not enabled. Exception for "emergency" users */
+        if (!isset($GLOBALS['enable_auditlog']) || !($GLOBALS['enable_auditlog'])) {
+            if ((soundex($user) != soundex("emergency")) && (soundex($user) != soundex("breakglass"))) {
```

Comment:
This soundex stuff just seems weird where we are just seeing if the user is in the `breakglass` aro, which should be done via the acl_get_group_titles($user) function in library/acl.inc library.

I am guessing it's gotta also slow things down a bit. Albeit, the acl_get_group_titles($user) may potentially be more expensive since will involve database calls.

Reply:
Notably, this same thing is done several more times below, so we should just do it once.

Reply:
Whatever we choose, would it make sense to do this once when the singleton is created?